java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -114,39 +113,63 @@ private void run(@Nullable final ExplodedGraph.Node currentNode, @Nullable final
-    Tree syntaxTree = parent.programPoint.syntaxTree();
-    if (syntaxTree != null && syntaxTree.is(Tree.Kind.METHOD_INVOCATION)) {
-      return currentNode.selectedMethodYield(parent);
+  private Stream<JavaFileScannerContext.Location> learnedConstraintFlow(@Nullable Constraint learnedConstraint, ExplodedGraph.Node currentNode, ExplodedGraph.Node parent) {
+    if (learnedConstraint == null || !addToFlow.test(learnedConstraint)) {
+      return Stream.empty();
+    }
+    Tree nodeTree = parent.programPoint.syntaxTree();
+    if (isMethodInvocationNode(parent)) {
+      return methodInvocationFlow(learnedConstraint, currentNode, parent);
     }
-    return null;
+    if (nodeTree.is(Tree.Kind.NEW_CLASS)) {
+      return Stream.of(location(parent, String.format("Constructor implies '%s'.", learnedConstraint.valueAsString())));
+    }
+    String name = SyntaxTreeNameFinder.getName(nodeTree);
+    String message = name == null ? learnedConstraint.valueAsString() : String.format("Implies '%s' is %s.", name, learnedConstraint.valueAsString());
+    return Stream.of(location(parent, message));
   }
 
-  private void flowForConstraint(ExplodedGraph.Node currentNode, ExplodedGraph.Node parent, @Nullable Constraint constraint, @Nullable MethodYield selectedMethodYield) {
-    if (constraint != null && addToFlow.test(constraint)) {
-      flow.add(location(parent, learnedConstraintMessage(constraint, currentNode, parent)));
-      if (selectedMethodYield != null) {
-        flow.addAll(flowFromMethodInvocation(selectedMethodYield, parent));
-      }
+  private Stream<JavaFileScannerContext.Location> methodInvocationFlow(Constraint learnedConstraint, ExplodedGraph.Node currentNode, ExplodedGraph.Node parent) {
+    MethodInvocationTree mit = (MethodInvocationTree) parent.programPoint.syntaxTree();
+    Stream.Builder<JavaFileScannerContext.Location> flowBuilder = Stream.builder();
+    if (currentNode.programState.peekValue() == symbolicValue) {
+      flowBuilder.add(location(parent, String.format("'%s()' returns %s.", mit.symbol().name(), learnedConstraint.valueAsString())));
+    }
+    SymbolicValue methodIdentifier = parent.programState.peekValues(mit.arguments().size() + 1).get(mit.arguments().size());
+    if (methodIdentifier == symbolicValue) {
+      flowBuilder.add(location(parent, "..."));
+    }
+    int argIdx = correspondingArgumentIndex(symbolicValue, parent);
+    if (argIdx != -1) {
+      ExpressionTree argTree = mit.arguments().get(argIdx);
+      String message = String.format("Implies '%s' is %s.", SyntaxTreeNameFinder.getName(argTree), learnedConstraint.valueAsString());
+      flowBuilder.add(new JavaFileScannerContext.Location(message, argTree));
     }
+    MethodYield selectedMethodYield = currentNode.selectedMethodYield(parent);
+    if (selectedMethodYield != null) {
+      selectedMethodYield.flow(argIdx).forEach(flowBuilder::add);
+    }
+    return flowBuilder.build();
   }
 
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)